### ISSUE TITLE
autosectionlabel generates smart-quoted labels only

### ISSUE THREAD

**Issue by wlonk** [sphinx-doc/sphinx#4027]
Subject: autosectionlabel generates smart-quoted labels only

### Problem
- The autosectionlabel extension is very useful, but some of my section labels include apostrophes. When I go to refer to them elsewhere in the text, I have to use smartquotes, despite those not being present in the source. I'd like to be able to refer to my autosectionlabels exactly as they are in the source.

#### Procedure to reproduce the problem
```
A sample title that's got an apostrophe
=======================================

This reference will not work: :ref:`A sample title that's got an apostrophe`.

This one will: :ref:`A sample title that’s got an apostrophe`.
```

I'm planning to put in a PR for this (at least behind a feature-flag), but I don't have the time right now, so this issue is a note-to-self, unless anyone wants to address it before I get a chance.

**Comment by jfbu** [comment]
This [docutils thread](https://sourceforge.net/p/docutils/mailman/message/35965815/) might be relevant (related: #3967). With now released Docutils 0.14, the rawsource data is preserved after Docutils smartquotification, which might possibly be relevant to the present issue.

**Comment by tk0miya** [comment]
Fixed by #4042 and #4591
